File: DSA/DSA_450/Python/10_stack_and_queues/027_Reverse_Queue.py
"""
Problem: Reverse a Queue
URL: https://practice.geeksforgeeks.org/problems/queue-reversal/1

Problem Statement:
Reverse all elements in a queue.

Sample Input/Output:
Input: queue [1,2,3,4,5]
Output: [5,4,3,2,1]
"""

from collections import deque


class Solution:
    def Reverse_Queue_Recursion(self, q):
        """
        Reverse queue using recursion.
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        if not q:
            return q
        front = q.popleft()
        reversed_q = self.Reverse_Queue_Recursion(q)
        reversed_q.append(front)
        return reversed_q

    def Reverse_Queue_Stack(self, q):
        """
        Reverse queue using stack.
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        st = []
        while q:
            st.append(q.popleft())
        while st:
            q.append(st.pop())
        return q


def Test_Reverse_Queue_Recursion():
    solution = Solution()
    q1 = deque([1, 2, 3, 4, 5])
    
    reversed1 = solution.Reverse_Queue_Recursion(q1)
    print("Recursion - Reversed Queue: ", end="")
    while reversed1:
        print(reversed1.popleft(), end=" ")
    print()


def Test_Reverse_Queue_Stack():
    solution = Solution()
    q2 = deque([1, 2, 3, 4, 5])
    
    reversed2 = solution.Reverse_Queue_Stack(q2)
    print("Stack - Reversed Queue: ", end="")
    while reversed2:
        print(reversed2.popleft(), end=" ")
    print()


if __name__ == "__main__":
    Test_Reverse_Queue_Recursion()
    Test_Reverse_Queue_Stack()

File: DSA/DSA_450/10_stack_and_queues/027_Reverse_Queue.cpp
/*
Problem: Reverse a Queue
URL: https://practice.geeksforgeeks.org/problems/queue-reversal/1

Problem Statement:
Reverse all elements in a queue.

Sample Input/Output:
Input: queue [1,2,3,4,5]
Output: [5,4,3,2,1]
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    queue<int> Reverse_Queue_Recursion(queue<int>& q) {
        if (q.empty()) return q;
        int front = q.front();
        q.pop();
        queue<int> reversed = Reverse_Queue_Recursion(q);
        reversed.push(front);
        return reversed;
    }

    queue<int> Reverse_Queue_Stack(queue<int>& q) {
        stack<int> st;
        while (!q.empty()) {
            st.push(q.front());
            q.pop();
        }
        while (!st.empty()) {
            q.push(st.top());
            st.pop();
        }
        return q;
    }
};

void Test_Reverse_Queue_Recursion() {
    Solution solution;
    queue<int> q1;
    q1.push(1);
    q1.push(2);
    q1.push(3);
    q1.push(4);
    q1.push(5);
    
    queue<int> reversed1 = solution.Reverse_Queue_Recursion(q1);
    cout << "Recursion - Reversed Queue: ";
    while (!reversed1.empty()) {
        cout << reversed1.front() << " ";
        reversed1.pop();
    }
    cout << endl;
}

void Test_Reverse_Queue_Stack() {
    Solution solution;
    queue<int> q2;
    q2.push(1);
    q2.push(2);
    q2.push(3);
    q2.push(4);
    q2.push(5);
    
    queue<int> reversed2 = solution.Reverse_Queue_Stack(q2);
    cout << "Stack - Reversed Queue: ";
    while (!reversed2.empty()) {
        cout << reversed2.front() << " ";
        reversed2.pop();
    }
    cout << endl;
}

int main() {
    Test_Reverse_Queue_Recursion();
    Test_Reverse_Queue_Stack();
    return 0;
}
